### PR TITLE
Feature/93 remove redundant api calls

### LINKED ISSUE #93: Fix redundant API Calls
Currently we have scripts that call the API, even though the relevant data is already stored in a local CSV file.
Especially when our datasets get bigger, this will cost SumUp unnecessary money.
Therefore we need an easy and efficient caching strategy, that can be used in all our data retrievers.

### Done?
* Retrieved API results are automatically stored in a local data cache (CSV file or maybe our database).
* Before new API calls are being made, a check should take place. If the requested data is already present locally, the API call is instead served from the local data cache.

### PR DESCRIPTION
**Pull request by felix-zailskas**
Closes #93 and #56 

Removed redundant API calls by implementing logic such that every step checks whether the fields it would add are already present. 

Every Step must now set the `added_cols` list. This should be a list of names of all the columns the step would add to the df. In the new method `Step.check_data_presence` it is checked if all of these columns are already present. If so the `Step.run` method is not executed. 
This behavior can be overridden by instantiating a step with the key word argument `Step(force_refresh=True)`.

Logger module was added. The custom logger allows logging at the levels
- debug
- info
- warning
- error
- critical

It logs to stdout and a file by default.

To use import the `get_logger()` function from the logger module and call it to get the logger object.

```
from logger import get_logger

log = get_logger()

log.info("This is information")
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -57,3 +56,9 @@ bin/
 
 # trained models
 **/models/*
+
+# # logs
+**/logs/*
+
+
+!.gitkeep
```
Comment (luccalb) [inline .gitignore:64]: is this really needed? .gitkeep file should be kept by default

File: src/bdc/steps/step.py [hunk 2]
```diff
@@ -34,6 +42,25 @@ def verify(self) -> bool:
         """
         raise NotImplementedError
 
+    def check_data_presence(self) -> bool:
```
Comment (luccalb) [inline src/bdc/steps/step.py:45]: Good Job!

File: tests/test_bdc_pipeline.py [hunk 3]
```diff
@@ -13,6 +13,7 @@
 
 class DummyStepOne(Step):
```
Comment (luccalb) [inline tests/test_bdc_pipeline.py:14]: Maybe add a testcase for checking the exisitng columns?

**Review by luccalb — CHANGES_REQUESTED** [review]
Great work! Thanks for taking care of the logging. Maybe just add one more test case.

**Review by rbbozkurt — APPROVED** [review]
Great job! And also  great patience to refactor the code to use log instead of print. However, is it possible to call log functions statically without calling the instance with `get_logger()`

File: .gitignore (hunk 1 above)
Comment (felix-zailskas) [reply .gitignore:64]: I had issues on my machine that the folder would not show up without this. But could be possible to remove it. Any idea what the harm could be leaving it in?

**Comment by felix-zailskas** [comment]
> Great job! And also great patience to refactor the code to use log instead of print. However, is it possible to call log functions statically without calling the instance with `get_logger()`

Right now it's not possible. The function get_logger() returns a singleton instance of the CustomLogger, which handles logging for the whole application.

File: .gitignore (hunk 1 above)
Comment (rbbozkurt) [reply .gitignore:64]: I do not think it harms anything, at most just double insurance for that files not be ignored at all.

Comment (felix-zailskas) [reply .gitignore:66]: I'll keep it then

File: tests/test_bdc_pipeline.py (hunk 3 above)
Comment (felix-zailskas) [reply tests/test_bdc_pipeline.py:14]: Done

File: tests/test_bdc_pipeline.py [hunk 4]
```diff
@@ -28,48 +52,91 @@ def finish(self) -> None:
...
-            self.p.run()
+            self.p_one.run()
 
-            # assert that all methods were run except for run()
+            # assert that all methods were run except for run() and check_data_presence()
             load_data_mock.assert_called_once()
             verify_mock.assert_called_once()
+            check_data_presence_mock.assert_called_once()
             run_mock.assert_not_called()
             finish_mock.assert_called_once()
```
Comment (luccalb) [inline tests/test_bdc_pipeline.py:116]: Nice work!

**Review by luccalb — APPROVED** [review]

**Pull request merged**